pennylane/qnn/keras.py

```diff
@@ -0,0 +1,292 @@
+            raise TypeError(
+                "QNode must include an argument with name {} for inputting data".format(
+                    self.input_arg
+                )
+            )
+
+        if self.input_arg in set(weight_shapes.keys()):
+            raise ValueError(
+                "{} argument should not have its dimension specified in "
+                "weight_shapes".format(self.input_arg)
+            )
+
+        if set(weight_shapes.keys()) | {self.input_arg} != set(self.sig.keys()):
+            raise ValueError("Must specify a shape for every non-input parameter in the QNode")
+
+        if qnode.func.var_pos:
+            raise TypeError("Cannot have a variable number of positional arguments")
+
+        if qnode.func.var_keyword:
+            raise TypeError("Cannot have a variable number of keyword arguments")
+
+        self.qnode = to_tf(qnode, dtype=tf.keras.backend.floatx())
+        self.weight_shapes = {
+            weight: (tuple(size) if isinstance(size, Iterable) else (size,) if size > 1 else ())
+            for weight, size in weight_shapes.items()
+        }
+        self.output_dim = output_dim[0] if isinstance(output_dim, Iterable) else output_dim
+
+        defaults = {
+            name for name, sig in self.sig.items() if sig.par.default != inspect.Parameter.empty
+        }
+        self.input_is_default = self.input_arg in defaults
+        if defaults - {self.input_arg} != set():
+            raise TypeError(
+                "Only the argument {} is permitted to have a default".format(self.input_arg)
+            )
+
+        self.weight_specs = weight_specs if weight_specs is not None else {}
+
+        self.qnode_weights = {}
+
+        super().__init__(dynamic=True, **kwargs)
+
+    def build(self, input_shape):
+        """Initializes the QNode weights.
+
+        Args:
+            input_shape (tuple or tf.TensorShape): shape of input data
+        """
+        for weight, size in self.weight_shapes.items():
+            spec = self.weight_specs.get(weight, {})
+            self.qnode_weights[weight] = self.add_weight(name=weight, shape=size, **spec)
+
+        super().build(input_shape)
+
+    def call(self, inputs):
+        """Evaluates the QNode on input data using the initialized weights.
+
+        Args:
+            inputs (tensor): data to be processed
+
+        Returns:
+            tensor: output data
+        """
+        outputs = []
+        for x in inputs:
+            qnode = self.qnode
+            for arg in self.sig:
+                if arg is not self.input_arg:
+                    w = self.qnode_weights[arg]
+                    qnode = functools.partial(qnode, w)
+                else:
+                    if self.input_is_default:
+                        qnode = functools.partial(qnode, **{self.input_arg: x})
+                    else:
+                        qnode = functools.partial(qnode, x)
+            outputs.append(qnode())
```

Comment:
Nitpick: I still think this method could be improved through the use of line comments explaining the logical flow

Reply:
Yes, your original comment had got lost! Have added these comments to the method now.